src/main/java/org/skriptlang/skript/bukkit/damagesource/DamageSourceWrapper.java

```diff
@@ -0,0 +1,137 @@
+package org.skriptlang.skript.bukkit.damagesource;
+
+import org.bukkit.Location;
+import org.bukkit.damage.DamageSource;
+import org.bukkit.damage.DamageType;
+import org.bukkit.entity.Entity;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Wrapper for continuous changing of a {@link DamageSource}.
+ */
+@SuppressWarnings("ALL")
+public class DamageSourceWrapper implements DamageSource {
+
+	private DamageType damageType = DamageType.GENERIC;
+	private @Nullable Entity causingEntity = null;
+	private @Nullable Entity directEntity = null;
+	private @Nullable Location damageLocation = null;
+	private @Nullable Location sourceLocation = null;
+	private boolean indirect = false;
+	private float foodExhaustion = 0f;
+	private boolean scalesWithDifficulty = false;
+
+	public DamageSourceWrapper() {}
+
+	@Override
+	public @NotNull DamageType getDamageType() {
+		return damageType;
+	}
+
+	@Override
+	public @Nullable Entity getCausingEntity() {
+		return causingEntity;
+	}
+
+	@Override
+	public @Nullable Entity getDirectEntity() {
+		return directEntity;
+	}
+
+	@Override
+	public @Nullable Location getDamageLocation() {
+		return damageLocation;
+	}
+
+	@Override
+	public @Nullable Location getSourceLocation() {
+		return sourceLocation;
+	}
+
+	@Override
+	public boolean isIndirect() {
+		return indirect;
+	}
+
+	@Override
+	public float getFoodExhaustion() {
+		return foodExhaustion;
+	}
+
+	@Override
+	public boolean scalesWithDifficulty() {
+		return scalesWithDifficulty;
+	}
+
+	public DamageSourceWrapper setDamageType(DamageType damageType) {
+		this.damageType = damageType;
+		return this;
+	}
+
+	public DamageSourceWrapper setCausingEntity(@Nullable Entity causingEntity) {
+		this.causingEntity = causingEntity;
+		return this;
+	}
+
+	public DamageSourceWrapper setDirectEntity(@Nullable Entity directEntity) {
+		this.directEntity = directEntity;
+		return this;
+	}
+
+	public DamageSourceWrapper setDamageLocation(@Nullable Location damageLocation) {
+		this.damageLocation = damageLocation;
+		return this;
+	}
+
+	public DamageSourceWrapper setSourceLocation(@Nullable Location sourceLocation) {
+		this.sourceLocation = sourceLocation;
+		return this;
+	}
+
+	public DamageSourceWrapper setIndirect(boolean indirect) {
+		this.indirect = indirect;
+		return this;
+	}
```

Comment:
this probably shouldn't be settable. indirect shouldn't be its own field, it should be dependent on the two entities.

Reply:
the documentation states "Damage is considered indirect if [getCausingEntity()](https://jd.papermc.io/paper/1.21.5/org/bukkit/damage/DamageSource.html#getCausingEntity()) is not equal to [getDirectEntity()](https://jd.papermc.io/paper/1.21.5/org/bukkit/damage/DamageSource.html#getDirectEntity())." so any implementation of DamageSource should adhere to that contract

Reply:
> as of yet atleast, hopefully.

Likely never will be ;(, too much is dependent on that stupid damage type